lib/charms/haproxy/v0/haproxy_route_tcp.py

```diff
@@ -0,0 +1,1738 @@
+    """
+
+    port: int = Field(description="The port exposed on the provider.", gt=0, le=65535)
+    backend_port: Optional[int] = Field(
+        description=(
+            "The port where the backend service is listening. " "Defaults to the provider port."
+        ),
+        default=None,
+        gt=0,
+        le=65525,
+    )
+    sni: Optional[VALIDSTR] = Field(
+        description=(
+            "Server name identification. Used to route traffic to the service. "
+            "Only available if TLS is enabled."
+        ),
+        default=None,
+    )
+    hosts: list[IPvAnyAddress] = Field(
+        description="The list of backend server addresses. Currently only support IP addresses.",
+        default=[],
+    )
+    check: Optional[TCPServerHealthCheck] = Field(
+        description="Configure health check for the service.",
+        default=None,
+    )
+    load_balancing: Optional[TCPLoadBalancingConfiguration] = Field(
+        description="Configure loadbalancing.", default=None
+    )
+    rate_limit: Optional[RateLimit] = Field(
+        description="Configure rate limit for the service.", default=None
+    )
+    bandwidth_limit: Optional[BandwidthLimit] = Field(
+        description="Configure bandwidth limit for the service.", default=None
+    )
+    retry: Optional[Retry] = Field(
+        description="Configure retry for incoming requests.", default=None
+    )
+    timeout: Optional[TimeoutConfiguration] = Field(
+        description="Configure timeout",
+        default=None,
+    )
+    server_maxconn: Optional[int] = Field(
+        description="Configure maximum connection per server", default=None
+    )
+    ip_deny_list: list[IPvAnyAddress] = Field(
+        description="List of IP addresses to block.", default=[]
+    )
+    enforce_tls: bool = Field(description="Whether to enforce TLS for all traffic.", default=True)
+    tls_terminate: bool = Field(description="Whether to enable tls termination.", default=True)
+
+    @model_validator(mode="after")
+    def assign_default_backend_port(self) -> "Self":
+        """Assign a default value to backend_port if not set.
+
+        The value is equal to the provider port.
+
+        Returns:
+            The model with backend_port default value applied.
+        """
+        if self.backend_port is None:
+            self.backend_port = self.port
+        return self
+
+    @model_validator(mode="after")
+    def sni_set_when_not_enforcing_tls(self) -> "Self":
+        """Check if sni is configured but TLS is disabled.
+
+        Raises:
+            ValueError: If sni is configured and TLS is disabled.
+
+        Returns:
+            The validated model.
+        """
+        if not self.enforce_tls and self.sni is not None:
+            raise ValueError("setting sni and disabling TLS are mutually exclusive.")
+        return self
+
+
+class HaproxyRouteTcpProviderAppData(_DatabagModel):
```

Comment:
Why wrapping this single property in a class?

Reply:
This is the representation of the application databag. It's declared as a class so that we can use the `from_json` method of the child class `_DatabagModel` to populate the class using the relation data regardless of how many attributes there are in the relation data. It's a convention that we set to uniformly manage relation data across the lib, wdyt?